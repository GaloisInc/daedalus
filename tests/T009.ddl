
def P =
  block
    x = UInt8
    y = UInt8

def Q =
  block
    let s = P
    $$    = ^s.x

def R =
  First
    a = UInt8
    b = Accept

def S =
  block
    let u = R
    $$ = u is a

