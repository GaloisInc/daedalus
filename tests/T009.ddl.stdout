module T009
 
--- Imports:
 
--- Type defs:
type T009.P = { x: uint 8
              ; y: uint 8
              }
 
type T009.R = Choose { b: {}
                     ; a: uint 8
                     }
 
--- Rules:
 
T009.P : Grammar T009.P =
  do (x : uint 8) <- Match UInt8
     (y : uint 8) <- Match UInt8
     pure {x = x,
           y = y}
 
T009.Q : Grammar (uint 8) =
  do (s : T009.P) <- T009.P
     ($$ : uint 8) <- pure (s.x)
     pure $$
 
T009.R : Grammar T009.R =
  Choose biased
    { {- a -} do (_10 : uint 8) <- Match UInt8
                 pure {a: _10}
    | {- b -} do (_11 : {}) <- pure {}
                 pure {b: _11}
    }
 
T009.S : Grammar (uint 8) =
  do (u : T009.R) <- T009.R
     ($$ : uint 8) <- case u is
                        { {| a = _12 |} -> pure _12
                        }
     pure $$
 
T009._P : Grammar {} =
  do @Match UInt8
     @Match UInt8
 
T009._Q : Grammar {} =
  T009._P
 
T009._R : Grammar {} =
  Choose biased
    { {- a -} @Match UInt8
    | {- b -} pure {}
    }
 
T009._S : Grammar {} =
  do (u : T009.R) <- T009.R
     case u is
       { {| a = _12 |} -> pure {}
       }
